### DISCUSSION TITLE
Splink 3.0 Large Data Recommendations

### DISCUSSION THREAD

**Discussion by MirandaLochner** [moj-analytical-services/splink#654] [Q&A] [answered]
Hi! I am looking to implement Splink 3.0 on a test dataset of 1.5 million record before scaling to 1 billion records. Are there any recommendations for running Splink on larger datasets?

I am using Spark and currently at 1.5 million I cannot get the job to complete for writing out the predictions DF. 

df_predictions = linker.predict(threshold_match_probability=0.95)
clusters = linker.cluster_pairwise_predictions_at_threshold(df_predictions, threshold_match_probability=0.95)

**Comment by ericmanning** [comment]
Hi Miranda --

Not a developer on this package, but I have some experience using it with large datasets.

The most likely issue here is that the blocking rules are too lenient, meaning the model is generating too many record comparisons. I found `count_num_comparisons_from_blocking_rule()` and `count_num_comparisons_from_blocking_rules_for_prediction()` useful for telling me how many comparisons my model will generate.

The problem is particularly acute for datasets with skewed counts in linkage fields. (For example, blocking on postcode and last name where you have N number of 'Smith's in one particularly large postcode will generate N^2 comparisons just for that zip/name combination.)

**Reply by MirandaLochner** [reply]
Thanks for the reply! Have you found an optimal count from applying the blocking comparison? Is 1 billion pairs from a 1.5 million row dataset too large?

**Reply by RobinL** [reply]
Generally, we try and keep the number of comparisons in the hundreds of millions.  But that's with input data of maybe 50 million records.   

1 billion input records is a very large number - it would be the largest use of Splink that I'm aware of.  You'd need:
- A fairly big Spark cluster
- Tight blocking rules

(the looser the blocking rules, the bigger the spark cluster you need, and there will probably be some upper limit to what is possible irrespective of cluster size)

If your data quality is high,  you can probably get away with tight blocking rules.

Note that for model training, you can use even tighter blocking rules, see [here](https://moj-analytical-services.github.io/splink/topic_guides/blocking_rules.html)

**Comment by MirandaLochner** [comment, marked as answer]
@RobinL  In addition to tight blocking rules, are there any tips for getting the linker.cluster_pairwise_predictions_at_threshold code to run faster/complete?

**Reply by RobinL** [reply]
I'll try and write up a proper topic guide on this, but if you play with some of the options on the Spark linker you can often get better performance

https://github.com/moj-analytical-services/splink/blob/014016695139e17209d8b9e7b037a02229e8b5c6/splink/spark/spark_linker.py#L67 the 'parquet' option here is probably fastest on large data, and try setting spark parallelism higher

**Reply by MirandaLochner** [reply]
Thanks for the advice! Parquet helped it complete the cluster step. Appreciate the tips, it helped the process run through 1.5 mill and 100 mill so I will continue to scale.

**Reply by RobinL** [reply]
I've now written a couple more topic guides that expand on the above answers a bit:

https://moj-analytical-services.github.io/splink/topic_guides/drivers_of_performance.html
https://moj-analytical-services.github.io/splink/topic_guides/optimising_spark.html
